### PR TITLE
SCP-2045: State machine produces thread tokens

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+module Plutus.Contract.StateMachine.ThreadToken where
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:15]: haddocks

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        minted = V.txInfoForge txinfo
+        expected = if burn
+            then inv (threadTokenValue ownSymbol vHash)
+            else      threadTokenValue ownSymbol vHash
+
+        -- True if the pending transaction mints the amount of
+        -- currency that we expect
+        mintOK =
+            let v = expected == minted
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:53]: this makes it impossible to mint any other currency that is unrelated to the thread token. You could use `assetClassValueOf` instead of comparing the entire `Value`

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 3]
```diff
@@ -312,10 +321,22 @@ runStep ::
     -- ^ The input to apply to the state machine
     -> Contract w schema e (TransitionResult state input)
 runStep smc input =
-    runGuardedStep smc input (\_ _ _ -> Nothing) >>= pure . \case
+    runGuardedStep smc input (\_ _ _ -> Nothing) <&> \case
         Left a  -> absurd a
         Right a -> a
 
+getThreadToken :: AsSMContractError e => Contract w schema e ThreadToken
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:328]: Missing haddocks

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 4]
```diff
@@ -331,10 +352,17 @@ runInitialise ::
     -- ^ The value locked by the contract at the beginning
     -> Contract w schema e state
 runInitialise StateMachineClient{scInstance} initialState initialValue = mapError (review _SMContractError) $ do
-    let StateMachineInstance{typedValidator, stateMachine} = scInstance
-        tx = mustPayToTheScript initialState (initialValue <> SM.threadTokenValue stateMachine)
-    let lookups = Constraints.typedValidatorLookups typedValidator
-    utx <- either (throwing _ConstraintResolutionError) pure (Constraints.mkTx lookups tx)
+    ownPK <- ownPubKey
+    utxo <- utxoAt (Ledger.pubKeyAddress ownPK)
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:356]: This works for now but in the future we should get rid of this as the `utxoAt` query generally returns too much data. Since we only need the bit that is specific to the `TxOutRef` of the thread token, maybe we could store this information in the thread token itself.

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs [hunk 5]
```diff
@@ -65,18 +64,23 @@ data StateMachine s i = StateMachine {
 
       -- | The 'AssetClass' of the thread token that identifies the contract
       --   instance.
-      smThreadToken :: Maybe AssetClass
+      smThreadToken :: Maybe TT.ThreadToken
```
Comment (j-mueller) [inline plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:67]: The comment could say that the initial step creates the thread token (if a thread token is defined)

**Review by j-mueller — APPROVED** [review]
This looks very good, great work! Just missing some haddocks

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 6]
```diff
@@ -312,10 +321,22 @@ runStep ::
...
         Left a  -> absurd a
         Right a -> a
 
+getThreadToken :: AsSMContractError e => Contract w schema e ThreadToken
+getThreadToken = mapError (review _SMContractError) $ do
+    ownPK <- ownPubKey
+    let constraints = mustPayToPubKey (pubKeyHash ownPK) (Ada.lovelaceValueOf 1)
+    utx <- either (throwing _ConstraintResolutionError) pure (Constraints.mkTx @Void mempty constraints)
+    tx <- balanceTx utx
+    txOutRef <-
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:334]: Is this all a fancy way of saying "get me an output belonging to the user, any output"?

Comment (michaelpj) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:334]: In particular I had to stare at it a bit because I was confused that the transaction wasn't getting spent :sweat_smile: 

This function probably needs a note saying that it latches on to a currently-unspent UTXO and you need to be careful in case it gets spent between you doing this and forging the token.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs [hunk 7]
```diff
@@ -331,10 +352,17 @@ runInitialise ::
...
-        tx = mustPayToTheScript initialState (initialValue <> SM.threadTokenValue stateMachine)
-    let lookups = Constraints.typedValidatorLookups typedValidator
-    utx <- either (throwing _ConstraintResolutionError) pure (Constraints.mkTx lookups tx)
+    ownPK <- ownPubKey
+    utxo <- utxoAt (Ledger.pubKeyAddress ownPK)
+    let StateMachineInstance{stateMachine, typedValidator} = scInstance
+        red = Ledger.Redeemer (PlutusTx.toData (Scripts.validatorHash typedValidator, False))
+        constraints = mustPayToTheScript initialState (initialValue <> SM.threadTokenValue scInstance)
+            <> mustMintValueWithRedeemer red (SM.threadTokenValue scInstance)
+            <> maybe mempty (mustSpendPubKeyOutput . ttOutRef) (smThreadToken stateMachine)
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine.hs:361]: worth some comments

Comment (michaelpj) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:361]: In particular, it looks like there's a failure case here (where you get the mempty out of the `maybe` call) that there isn't in the line above. Can those come apart? Similarly when you destroy the token.

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs [hunk 8]
```diff
@@ -65,18 +64,23 @@ data StateMachine s i = StateMachine {
 
       -- | The 'AssetClass' of the thread token that identifies the contract
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:65]: out of date

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs [hunk 9]
```diff
@@ -65,18 +64,23 @@ data StateMachine s i = StateMachine {
...
+      smThreadToken :: Maybe TT.ThreadToken
     }
 
-{-# INLINABLE threadTokenValue #-}
+{-# INLINABLE threadTokenValueInner #-}
 -- | The 'Value' containing exactly the thread token, if one has been specified.
-threadTokenValue :: StateMachine s i -> Value
-threadTokenValue StateMachine{smThreadToken} = maybe mempty (\c -> Value.assetClassValue c 1) smThreadToken
+threadTokenValueInner :: Maybe TT.ThreadToken -> ValidatorHash -> Value
+threadTokenValueInner = maybe (const mempty) (TT.threadTokenValue . TT.ttCurrencySymbol)
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:73]: am I the only person who likes pattern matching any more? :p

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+data ThreadToken = ThreadToken
+    { ttOutRef         :: TxOutRef
+    , ttCurrencySymbol :: CurrencySymbol
+    }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.makeLift ''ThreadToken
+
+validate :: TxOutRef -> (ValidatorHash, Bool) -> V.ScriptContext -> Bool
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:40]: I know it's a trivial usecase, but it would make me feel less like I had to audit all the use sites if we took `data MintingPolarity = Mint | Burn` or something.

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            let v = expected == minted
+            in traceIfFalse "Value minted different from expected" v
+
+        -- True if the pending transaction spends the output
+        -- identified by @(refHash, refIdx)@
+        txOutputSpent =
+            let v = V.spendsOutput txinfo refHash refIdx
+            in  traceIfFalse "Pending transaction does not spend the designated transaction output" v
+
+    in mintOK && (burn || txOutputSpent)
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:62]: I'd clarify that this is logically equivalent to `mint => txOutputSpent` which is what we want.

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+curPolicy :: TxOutRef -> MintingPolicy
+curPolicy outRef = mkMintingPolicyScript $
+    $$(PlutusTx.compile [|| \r -> Scripts.wrapMintingPolicy (validate r) ||])
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:66]: is this... actually using a redeemer?? :exploding_head: 

nice

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    { ttOutRef         :: TxOutRef
+    , ttCurrencySymbol :: CurrencySymbol
+    }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.makeLift ''ThreadToken
+
+validate :: TxOutRef -> (ValidatorHash, Bool) -> V.ScriptContext -> Bool
+validate (TxOutRef refHash refIdx) (vHash, burn) ctx@V.ScriptContext{V.scriptContextTxInfo=txinfo} =
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:41]: Doesn't this let you burn the token at any time? That seems too strong. Someone could burn the token in the middle of the machine operation. Something needs to keep it from being burned until the end. Although maybe that's handled by the state machine's balancing operations? Seems subtle, and worth a note. TBH, the whole setup is worth a note explaining how it all fits together!

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        `PlutusTx.applyCode`
+            PlutusTx.liftCode outRef
+
+{-# INLINABLE threadTokenValue #-}
+-- | The 'Value' containing exactly the thread token.
+threadTokenValue :: CurrencySymbol -> ValidatorHash -> Value
+threadTokenValue currency (ValidatorHash vHash) = Value.singleton currency (TokenName vHash) 1
+
+{-# INLINABLE checkThreadToken #-}
+checkThreadToken :: Maybe ThreadToken -> ValidatorHash -> Value -> Bool
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs:76]: What does this do? It looks like it not only checks that `threadTokenValue` is less than the given value, but in fact checks that it's the only token of this currency? why?

**Review by michaelpj — APPROVED** [review]
Looks good, but could really do with more explanatory comments. Lots of this is quite subtle, we should overdo the explanations if anything!

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs (hunk 9 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:73]: 😆 
I wanted to write it as `foldMap (TT.threadTokenValue . TT.ttCurrencySymbol)` but the Plutus Monoid class doesn't have an instance for functions.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 4 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs:353]: I have added TODO comments

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 7 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs]: I have tried to do this a bit, but constraints are not completely composable. For example, I tried splitting `mustPayToTheScript initialState (initialValue <> SM.threadTokenValue scInstance)` into separate `mustPayToTheScript initialState initialValue` and `mustPayToTheScript initialState (SM.threadTokenValue scInstance)` but that broke tests. (I didn't investigate further)

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs (hunk 13 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs]: This one I'm really not sure about. I don't feel at home enough with all of this yet to be able to think it all through. @j-mueller what do you think?

Comment (michaelpj) [reply plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs]: Could you write down what you *think* happens? Then we can see if we all agree.

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 7 above)
Comment (michaelpj) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs]: I'm not saying it's wrong, I just want an argument explaining why it isn't wrong!

File: plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs (hunk 13 above)
Comment (j-mueller) [reply plutus-contract/src/Plutus/Contract/StateMachine/ThreadToken.hs]: The state machine validator prevents burning the token in the middle of the machine operation, even if the MPS would allow it (so you can't spend the state machine output to get the token out)

File: plutus-contract/src/Plutus/Contract/StateMachine.hs (hunk 7 above)
Comment (j-mueller) [reply plutus-contract/src/Plutus/Contract/StateMachine.hs]: Merging two `mustPayToTheScript` constraints is tricky (what if they have different datum values? What if the user actually wants to create two outputs?)

But I see that @sjoerdvisscher changed it to create the minting constraint and the must spend pubkey output constraint together, or not at all. So it's better now.

**Comment by j-mueller** [comment]
Alright I think we can merge this!
